File: data_structure/union_find.cc
#include <iostream>
#include <vector>

class UnionFind
{
public:
    std::vector<int> data;

    UnionFind() = delete;
    explicit UnionFind(int size) :data(size, -1) {}

    bool unionSet(int x, int y) {
        x = root(x); y = root(y);
        if (x != y) {
            if (data[y] < data[x]) std::swap(x, y);
            data[x] += data[y]; data[y] = x;
        }
        return x != y;
    }

    bool findSet(int x, int y) {
        return root(x) == root(y);
    }

    int root(int x) {
        return data[x] < 0 ? x : data[x] = root(data[x]);
    }

    int size(int x) {
        return -data[root(x)];
    }
};

int main(void)
{
    UnionFind uf(10);
    uf.unionSet(1, 2);
    uf.unionSet(3, 4);
    uf.unionSet(1, 5);

    std::cout << uf.size(1) << std::endl;
    std::cout << uf.root(5) << std::endl;
    return 0;
}

File: data_structure/union_find/union_find.py

class UnionFind():
    def __init__(self, size):
        self.data = [-1] * size

    def union_set(self, x, y):
        x = self.root(x)
        y = self.root(y)

        if (x != y):
            if (self.data[y] < self.data[x]):
                x, y = y, x

            self.data[x] += self.data[y]
            self.data[y] = x

        return x != y

    def find_set(self, x, y):
        return self.root(x) == self.root(y)

    def root(self, x):
        if self.data[x] < 0:
            return x
        else:
            self.data[x] = self.root(self.data[x])
            return self.data[x]

    def size(self, x):
        return -self.data[self.root(x)]


if __name__ == "__main__":
    uf = UnionFind(10)

    uf.union_set(1, 2)
    uf.union_set(2, 3)

    print(uf.root(3))
